src/data_provider/src/extended_sources/users/src/shadow_linux.cpp

```diff
@@ -0,0 +1,100 @@
+/* Copyright (C) 2015, Wazuh Inc.
+ * All rights reserved.
+ *
+ * This program is free software; you can redistribute it
+ * and/or modify it under the terms of the GNU General Public
+ * License (version 2) as published by the FSF - Free Software
+ * Foundation.
+ */
+
+#include <iostream>
+#include <regex>
+
+#include "shadow_wrapper.hpp"
+#include <shadow_linux.hpp>
+
+ShadowProvider::ShadowProvider(std::shared_ptr<IShadowWrapper> shadowWrapper)
+    : m_shadowWrapper(std::move(shadowWrapper))
+{
+}
+
+ShadowProvider::ShadowProvider()
+    : m_shadowWrapper(std::make_shared<ShadowWrapper>())
+{
+}
+
+nlohmann::json ShadowProvider::collect()
+{
+    nlohmann::json results = nlohmann::json::array();
+    struct spwd* shadow_entry;
+    const auto kPasswordHashAlgRegex = std::regex("^\\$(\\w+)\\$");
+
+    // Acquire exclusive access to the shadow file
+    if (m_shadowWrapper->lckpwdf() == -1)
+    {
+        //TODO: Logs and Error with "lckpwdf"
+        return results;
+    }
+
+    // Rewind the shadow file to the beginning
+    m_shadowWrapper->setspent();
+
+    // Read the first shadow password entry
+    while ((shadow_entry = m_shadowWrapper->getspent()) != NULL)
+    {
+        nlohmann::json entry;
+        entry["last_change"] = shadow_entry->sp_lstchg;
+        entry["min"] = shadow_entry->sp_min;
+        entry["max"] = shadow_entry->sp_max;
+        entry["warning"] = shadow_entry->sp_warn;
+        entry["inactive"] = shadow_entry->sp_inact;
+        entry["expire"] = shadow_entry->sp_expire;
+        entry["username"] = shadow_entry->sp_namp != nullptr ? shadow_entry->sp_namp : "";
+        // sp_flag - reserved for future use, won't be added.
+        // entry["flag"] = shadow_entry->sp_flag;
+
+        if (shadow_entry->sp_pwdp != nullptr)
+        {
+            std::string password = std::string(shadow_entry->sp_pwdp);
+            std::smatch matches;
+
+            if (password == "!!")
+            {
+                entry["password_status"] = "not_set";
+            }
+            else if (password[0] == '!' || password[0] == '*' || password[0] == 'x')
+            {
+                entry["password_status"] = "locked";
+            }
+            else if (password.empty())
+            {
+                entry["password_status"] = "empty";
+            }
+            else
+            {
+                entry["password_status"] = "active";
+            }
+
+            if (std::regex_search(password, matches, kPasswordHashAlgRegex))
```

Comment:
Just for keep in mind, it may be useful to include the name of the algorithm in the future.

```c++
static const std::unordered_map<std::string, std::string> kHashIdToName = {
    {"1", "md5"},
    {"2a", "bcrypt"},
    {"2b", "bcrypt"},
    {"2y", "bcrypt"},
    {"3", "nt-hash"},
    {"5", "sha256"},
    {"6", "sha512"}
};

```

Reply:
As I mentioned, I only intend to leave a reference for the future.